shared-module/displayio/OnDiskBitmap.c

```diff
@@ -56,42 +56,65 @@ void common_hal_displayio_ondiskbitmap_construct(displayio_ondiskbitmap_t *self,
     uint16_t bits_per_pixel = bmp_header[14];
     uint32_t compression = read_word(bmp_header, 15);
     uint32_t number_of_colors = read_word(bmp_header, 23);
+
+    bool indexed = ((bits_per_pixel <= 8) && (number_of_colors != 0));
     self->bitfield_compressed  = (compression == 3);
-    
-    self->grayscale = ((bits_per_pixel == 8) && (number_of_colors == 256));
+    self->bits_per_pixel = bits_per_pixel;
+    self->width = read_word(bmp_header, 9);
+    self->height = read_word(bmp_header, 11);
+
     if (bits_per_pixel == 16){
-        if (((header_size == 124) || (header_size == 56)) && (self->bitfield_compressed)) {
+        if (((header_size >= 56)) || (self->bitfield_compressed)) {
             self->r_bitmask = read_word(bmp_header, 27);
             self->g_bitmask = read_word(bmp_header, 29);
             self->b_bitmask = read_word(bmp_header, 31);
             
-            if (!((self->r_bitmask == 0xf800) && (self->g_bitmask == 0x07e0) && (self->b_bitmask == 0x001f))){
-                mp_raise_ValueError_varg(translate("Only 16bpp RGB 565 supported for bitfield compressed BMPs;  R:%x G:%x B:%x"), 
-                self->r_bitmask, self->g_bitmask, self->b_bitmask);
-            }
-        } else if (header_size == 40){ // no bitmasks means 5:5:5
+        } else { // no compression or short header means 5:5:5
             self->r_bitmask = 0x7c00;
             self->g_bitmask = 0x3e0;
             self->b_bitmask = 0x1f;
         }
+    } else if ((indexed) && (self->bits_per_pixel != 1)) {
+        uint16_t palette_size = number_of_colors * sizeof(uint32_t);
+        uint16_t palette_offset = 0xe + header_size;
+
+        self->palette_data = m_malloc(palette_size, false);
+
+        f_rewind(&self->file->fp);
+        f_lseek(&self->file->fp, palette_offset);
+
+        UINT palette_bytes_read;
+        if (f_read(&self->file->fp, self->palette_data, palette_size, &palette_bytes_read) != FR_OK) {
+            mp_raise_OSError(MP_EIO);
+        }
+        if (palette_bytes_read != palette_size) {
+            mp_raise_ValueError(translate("Unable to read color palette data"));
+        }
+
 
-    } else if (!(header_size == 12 || header_size == 40 || header_size == 108 || header_size == 124) ||
-        !(compression == 0)) {
+    } else if (!(header_size == 12 || header_size == 40 || header_size == 108 || header_size == 124)) {
         mp_raise_ValueError_varg(translate("Only Windows format, uncompressed BMP supported %d"), header_size);
     }
 
-    if (bits_per_pixel < 16 && !(self->grayscale)) {
-        mp_raise_ValueError_varg(translate("Only true color (24 bpp or higher), 16bpp 565 and 555, and 8bpp grayscale BMP supported %d"), bits_per_pixel);
+    if ((bits_per_pixel == 4 ) || (( bits_per_pixel == 8) && (number_of_colors == 0))) {
+        mp_raise_ValueError_varg(translate("Only monochrome, indexed 8bpp, and 16bpp or greater BMPs supported: %d"), bits_per_pixel);
```

Comment:
Can you change this message to be clearer about what the printed numeric value is, like "...supported, not %d bpp" or something like that.